apps/shipments/serializers.py

```diff
@@ -377,3 +379,86 @@ class Meta:
 
     def create(self, validated_data):
         return TrackingData.objects.create(**validated_data, **self.context)
+
+
+class ChangesDiffSerializer:
+    def __init__(self, queryset):
+        self.queryset = queryset
+
+        self.relation_fields = ('ship_from_location', 'ship_to_location', 'final_destination_location',
+                                'bill_to_location', )
+
+        self.excluded_fields = ('history_user', )
+
+    def diff_object_fields(self, old, new):
+        changes = new.diff(old)
+
+        flat_changes = self.build_list_changes(changes)
+
+        relation_changes = self.relation_changes(new)
+
+        return {
+            'history_date': new.history_date,
+            'fields': flat_changes,
+            'relationships': relation_changes if relation_changes else None,
+            'author': new.history_user,
+        }
+
+    def build_list_changes(self, changes):
+        field_list = []
+        if changes is None:
+            return field_list
+
+        changes_list = [change for change in changes.changes if change.field not in self.excluded_fields]
+        for change in changes_list:
+            field = {
+                'field': change.field,
+                'old': change.old,
+                'new': change.new
+            }
+
+            if change.field in self.relation_fields:
+                if change.old is None:
+                    field['new'] = Location.history.filter(pk=change.new).first().instance.id
+                if change.old and change.new:
+                    continue
+
+            field_list.append(field)
+        return field_list
+
+    def relation_changes(self, new_historical):
+
+        relations_map = {}
+        for relation in self.relation_fields:
+            historical_relation = getattr(new_historical, relation, None)
+            changes = None
+            if historical_relation:
+                date_max = new_historical.history_date
+                date_min = date_max + timedelta(milliseconds=-500)
```

Comment:
```suggestion
                date_min = date_max - timedelta(milliseconds=500)
```
Where does the 500ms window come from? Are we sure that this window is always going to work in every case?

Reply:
On my local environment, I did observe a fluctuation around `200 ms` I arbitrary choose the 500 ms to have some confidence on the real server . But we can always tweak it the way we want. Will you be ok with that value set up as an environment variable so that you can adjust it in case we found a wider discrepancy on the server?

Reply:
Yeah I think it makes sense to make it a setting, something like `SIMPLE_HISTORY_RELATED_WINDOW_MS` and set it to 1000 for now.